### ISSUE TITLE
Directory.EnumerateFileSystemEntries doesn't return any item when specify pattern like with "[]"

### ISSUE THREAD

**Issue by EmmaZhu** [dotnet/corefx#20688]
`Directory.EnumerateFileSystemEntries` doesn't return items when search pattern include "[...]" in it on Ubuntu (It works well on Windows, while I didn't test it on other platforms). 
For example, I have a file named "[a]" under current folder, following code doesn't output anything:

```
         foreach (var fileItem in Directory.EnumerateFileSystemEntries(".", "[a]", SearchOption.TopDirectoryOnly))
            {
                Console.WriteLine(fileItem);
            }
```

Anyone could help to take a look at this?

**Comment by JeremyKuhne** [comment]
@EmmaZhu that should fail, shouldn't it? You're asking for any files of the set [a], which should be only "a". Did you try "\\[a\\]"?

**Comment by JeremyKuhne** [comment]
Looked into this a bit more. Mono is trying to match Windows behavior, and we are as well, but we're failing to account for this. I'll change this to escape `[`.  It also looks like we need to escape backslash. (Mono does their own matching, even though they mention `fnmatch(3)`.)

If we really want this to be correct we should also look at correctly handling `?`. Not sure if this is possible with [`FNM_EXTMATCH`](http://man7.org/linux/man-pages/man3/fnmatch.3.html)? The real rule is that question marks are optional if they precede a period or end-of-string. So `foo???.txt` matches any filename that starts with `foo`, has up to 3 more characters, and has the extension `.txt`. If I'm reading the docs right, that would be something like `foo?(?|??|???).txt`??

Additionally a trailing period is optional _if_ it follows a `*` or `?`. I've documented the full [Windows behavior](https://blogs.msdn.microsoft.com/jeremykuhne/2017/06/04/wildcards-in-windows/) (possibly for the first time ever).
 
@stephentoub, @ianhays, thoughts?

**Comment by stephentoub** [comment]
I'm unclear... what is the extent of the cases that aren't "correctly" handled today, and what is the impact of not handling them in the prescribed manner?

And what is the proposed fix?

**Comment by ianhays** [comment]
>Looked into this a bit more. Mono is trying to match Windows behavior, and we are as well, but we're failing to account for this. I'll change this to escape [. It also looks like we need to escape backslash. (Mono does their own matching, even though they mention fnmatch(3).)

I suppose I prefer maintaining consistency with mono/windows, though I'm on the fence about it. Would a Unix dev be familiar with this weird Windows behavior that is based on historical decisions?

**Comment by JeremyKuhne** [comment]
> I'm unclear... what is the extent of the cases that aren't "correctly" handled today, and what is the impact of not handling them in the prescribed manner?

- Using a `[` in your search pattern won't work consistently across Windows/Unix (see the user's scenario at the top)
- Having a `\` in your search pattern on Unix escapes, on Windows it doesn't (not a valid char on Windows, but you won't be able to search for files on Unix without knowing that you have to double escape)
- Searching for `Foo???.txt` will return different results on Unix/Windows
- Searching for `Foo?.` and `Foo*.` will give different results on Unix/Windows

It is bad news for Azure scripting any way you slice it (hence the source of the issue I assume).

> And what is the proposed fix?

(1) The easy part is to escape `\` and `[` to `\\` and `\[\` when searching. That would be pri-0 I think.

(2) Not quite-as-easy are `?` runs before periods and end-of-string. When they are there they are `.{0,n}` in regex-speak. So `Foo???.txt` would be `Foo.{0,3}\.txt` as a regex. Not sure if something equivalent is possible with `fnmatch(3)`.

(3) Periods are even trickier- my regex-fu is lacking. If they are before a `*` or `?` they don't have to match **if** they are at the end of the string. So `foo.???` will match anything without an extension or up to a three character extension. 🙈 

I want to plow forward with separate fixes for these as there is diminishing return and increasing complexity. Being consistent is important here- unfortunately our consistency source is nearly 30 years old (over 50 if you count what it was trying to emulate).

I'm planning to follow up with a new feature to allow matching in  "modern" styles. Perhaps multiple options from straight (no caveats) `*` & `?`, up to full-blown regex.

**Comment by danmosemsft** [comment]
@JeremyKuhne I guess we can close this now?

**Comment by danmosemsft** [comment]
@JeremyKuhne I'm closing this because it seems all fixed or tracked separately.

**Comment by JeremyKuhne** [comment]
Yes, this is resolved by the linked PRs (#20784 & #20822)
